source/fe/mapping_q_cache.cc

```diff
@@ -125,6 +129,97 @@ MappingQCache<dim, spacedim>::initialize(
 
 
 
+template <int dim, int spacedim>
+void
+MappingQCache<dim, spacedim>::initialize(
+  const Triangulation<dim, spacedim> &tria,
+  const Mapping<dim, spacedim> &      mapping,
+  const std::function<Point<spacedim>(
+    const typename Triangulation<dim, spacedim>::cell_iterator &,
+    const Point<spacedim> &)> &       transformation_function,
+  const bool                          is_displacement_function)
+{
+  // FE and FEValues in the case they are needed
+  FE_Nothing<dim, spacedim> fe;
+  Threads::ThreadLocalStorage<std::unique_ptr<FEValues<dim, spacedim>>>
+    fe_values_all;
+
+  this->initialize(
+    tria,
+    [&](const typename Triangulation<dim, spacedim>::cell_iterator &cell) {
+      std::vector<Point<spacedim>> points;
+
+      if (const auto m =
+            dynamic_cast<const MappingQGeneric<dim, spacedim> *>(&mapping))
+        {
+          points = m->compute_mapping_support_points(cell);
+        }
+      else
+        {
+          // get FEValues (thread-safe); in the case that this thread has not
+          // created a an FEValues object yet, this helper-function also
+          // creates one with the right quadrature rule
+          auto &fe_values = [&]() -> FEValues<dim, spacedim> & {
+            auto &fe_values = fe_values_all.get();
+            if (fe_values.get() == nullptr)
```

Comment:
This seems like a gratuitous use of a lambda function :-) The only thing you do is immediately call it, then assign its result to a reference, and then call two functions on the reference. You might as well duplicate the two calls into the `if` and `else` branches -- the overall number of lines is going to be less than what you have right now, and I bet it will be easier to read :-)